coderd/provisionerdserver/provisionerdserver.go

```diff
@@ -1954,14 +1954,27 @@ func (s *server) completeWorkspaceBuildJob(ctx context.Context, job database.Pro
 		}
 
 		appIDs := make([]string, 0)
+		agentIDByAppID := make(map[string]uuid.UUID)
 		agentTimeouts := make(map[time.Duration]bool) // A set of agent timeouts.
 		// This could be a bulk insert to improve performance.
 		for _, protoResource := range jobType.WorkspaceBuild.Resources {
-			for _, protoAgent := range protoResource.Agents {
+			for ai, protoAgent := range protoResource.Agents {
+				if protoAgent == nil {
+					continue
+				}
+				// Always set the agent ID. Previously this ID was generated
+				// in InsertWorkspaceResource, but it's required for linking
+				// the agent to a task. For now we overwrite the
+				// protoAgent.Id because that matches old behavior.
+				agentID := uuid.New()
+				protoAgent.Id = agentID.String()
```

Comment:
Review: This is potentially the most controversial change. I've retain previous behavior in `InsertWorkspaceResource` if this value isn't set (e.g. template import).

This basically moves the behavior one level up so that we can access the ID here.